### ISSUE TITLE
SMTP Authentication Credentials before Sending

### ISSUE THREAD

**Issue by thabaum** [dnnsoftware/Dnn.Platform#3121]
```
public static void CreateTestMessage1(string server, int port)
{
    string to = "jane@contoso.com";
    string from = "ben@contoso.com";
    string subject = "Using the new SMTP client.";
    string body = @"Using this new feature, you can send an email message from an application very easily.";
    MailMessage message = new MailMessage(from, to, subject, body);
    SmtpClient client = new SmtpClient(server, port);
    // Credentials are necessary if the server requires the client 
    // to authenticate before it will send email on the client's behalf.
    **client.Credentials = CredentialCache.DefaultNetworkCredentials;**
    client.Send(message);
}
```

According to this example **client.Credentials = CredentialCache.DefaultNetworkCredentials;** should be used to authenticate before it sends email.  

How can we make DNN CE do this as an option that you can set a check box for that would push this to the server prior to trying to send the email?

I cannot send email from a server due to being unable to authenticate first on my SSL 465 port.  The TLS needs STARTTLS pushed to the mail server so the 587 port is a no go as well.

Regardless SSL=TLS and I would at least like to be able to use SSL port but I need to be able to authenticate the credentials prior to sending mail.

I believe all of the logic is in this file in dnn is here:  https://github.com/dnnsoftware/Dnn.Platform/blob/development/DNN%20Platform/DotNetNuke.Log4net/log4net/Appender/SmtpAppender.cs

Any ideas on where that line should go and how it should look?

**Comment by mitchelsellers** [comment]
This is already supported within DNN when using the NTLM setting.  It is using a slightly different method, but it does the same thing: https://github.com/dnnsoftware/Dnn.Platform/blob/development/DNN%20Platform/Library/Services/Mail/Mail.cs#L148

**Comment by thabaum** [comment]
I believe I mentioned this in my other threed 3008.  It uses NTLM but that is what you use for basic authentication to authenticate before sending... at least that is what I believe it should do so you add the NTLM line above the Smtp send message and it should be the method ran for authentication FIRST prior to sending... simple

put a check box to allow authentication prior to sending put that line for that option in the same one... pretty much combine the two options.

Isnt that what it is saying in the example from microsoft.  I was trying to find why DNN does not authenticate first... this is my first clue. what do you have for the answer if it isnt? close the thread?

**Comment by thabaum** [comment]
> `                        smtpClient.EnableSsl = smtpEnableSSL;
client.Credentials = CredentialCache.DefaultNetworkCredentials;
>                         smtpClient.Send(mailMessage);
>                         smtpClient.Dispose();`

**Comment by thabaum** [comment]
but i think the defaultnetworkcredentials might be replaced with the SMTP credentials collected but they are cached?

**Comment by mitchelsellers** [comment]
NTLM != Basic Authentication.

Basic Authentication = Uses the provided username password
NTLM = Uses the defaults as mentioned above.

**Comment by thabaum** [comment]
Basic Authentication + Authenticate Prior to Sending emails which is shown in the example i posted in original thread 

CredentialCache.

is what I am looking for in the DNN code I dont see it.  This would be another option to set as not all servers need this to authenticate and send.   Mine does.

**Comment by thabaum** [comment]
![image](https://user-images.githubusercontent.com/13577556/66284393-5ae42280-e87c-11e9-89c6-f9f97831f542.png)

**Comment by mitchelsellers** [comment]
There is no such thing as Authentication Before Send, that is simply authentication on the connection.

Without knowing the in's and outs of your configuration, this appears to be a configuration that within DNN should be using the Basic authentication open, as you are asking to use a normal credentialed account.

We don't support any form of a credential cache, aside from the stored username/password for basic authentication.

**Comment by thabaum** [comment]
![image](https://user-images.githubusercontent.com/13577556/66284643-a814c400-e87d-11e9-9204-cd51585335ad.png)

**Comment by thabaum** [comment]
> We don't support any form of a credential cache, aside from the stored username/password for basic authentication.

I suppose this is a feature request...  I am looking for community support not "DNN Corp" support on this one... i gave up on that, that ship sailed few years back i believe.  But at least we are all on the same page now...

**Comment by mitchelsellers** [comment]
I guess I would need to see your configuration DNN & Mail Server, this isn't the venue to try and fix due to privacy

**Comment by thabaum** [comment]
I agree not posting more than what you see.  And I am not really into sharing either.

What I can do is work with you to setup a test environment.  I will email later if interested.

**Comment by mitchelsellers** [comment]
I don't have the time to try and configure my own mail server.  We know that the following configurations work with the current DNN Installation/features.

* Local SMTP
* Office 365 SMTP
* Hosted Exchange SMTP
* SendGrid & MailGun SMTP
* Various other third-party hostings
So this would most likely be a very specific config.

**Comment by thabaum** [comment]
I can setup an environment as we can probably just use one I already have setup.  Or I can set one up quickly, it is not exchange but I can set those up as well if needed... obviously not hosted exchange or office 365....

I can get it to work locally that is my work around sadly I have to have my mail server on my IIS server... I am trying to seperate.

**Comment by valadas** [comment]
Ok, I have my mail servers setup a bit like this where they require authetication prior to sending mail out. You can use basic authentication with or without ssl (tls).
What you see in Outlook settings to login with existing credentials, behind the curtains it just uses your incoming settings login/password. Dnn does not have that because Dnn only sends and does not receive, so it would not know your login/password from the incoming settings like Outlook does...

**Comment by thabaum** [comment]
I was just wondering why I am getting a couple errors I suppose and a another one .  I have had a tough time getting my mail server and DNN to like eachother for some reason. If it is working for you with authentication prior to sending emails then it should work for me.  I will test out my environment further.   I am getting two errors when connecting depending if I connect DNN  with or without authentication:

**Message:Unable to read data from the transport connection: net_io_connectionclosed.
and 
Message:User not local; please try a different path. The server response was: This mail server requires authentication before sending mail from a locally hosted domain. Please reconfigure your mail client to authenticate before sending mail.**

I look up the second error and I find threads like this:

https://stackoverflow.com/questions/38692563/error-unable-to-read-data-from-the-transport-connection-net-io-connectionclosed

This reference is a similar situation I believe to what I am dealing with potentially here with the STARTTLS.  The answer on this thread is where my mind is I believe on this one and trying to figure out how to deal with it and again maybe just leads to a library like MailKit being a solution.

My mail server only supports TLS (SSL) version 1.3 if that makes any difference.  TLS 1.2 and lower are not accepted connections.  And STARTTLS is supported feature and to use the port you must use SSL or TLS or your connection is dropped.

I have one port legacy setup with SSL... and the other setup strictly for TLS only.  All mail ports should support TLS with the STARTTLS command

telnet IP 25
ehlo example.com
...
250 STARTTLS

To dig any deeper I really need to discuss in private.  I will email if I can figure it out if it is only related to my server and no one else I may just have to deal with this one alone.

If these errors ring any bells for others that dealt with this please let me know I would greatly appreciate it.

 Thank you for verifying it currently works for you with similar configurations.  I will reset passwords again and see if I can clear up any network connectivity issues if I can find any.

**Comment by valadas** [comment]
Ok so for this one:

> Message:User not local; please try a different path. The server response was: This mail server requires authentication before sending mail from a locally hosted domain. Please reconfigure your mail client to authenticate before sending mail.

This is probably when you connect with no credentials and your server requires to login to send out, this is good unless you have a very specific setup with other protections in place. So I would think using the "Basic authentication" together with SSL would be the settings you need here.

> My mail server only supports TLS (SSL) version 1.3 if that makes any difference.
This may be a problem if it is true. Don't quote me on this, but I am not sure Microsoft adpted TLS1.3 yet, that just got released in 2018 and only a handful of browsers support it on their latest release. So if you are sure bout this, I would start by trying to get TLS 1.2 enabled on that mail server and make sure the web server is updated to latests patches, I think TLS 1.2 only came out of the box on Windows Server 2012 or some such... See https://docs.microsoft.com/en-us/windows/win32/secauthn/protocols-in-tls-ssl--schannel-ssp-

And another "don't quote me on this" thing. SSL and TLS is often used interchangebly and it is not a choice made by the sender of the emails, the os does a handshake with the other socket and they decide together how to secure the connection, this part is not specific to emails, it is just a TCP/IP think no matter what you are transfering. In other words, SSL/TLS is application idenpendent.

And a final "don't quote me on this", STARTTLS is oportunistic TLS, it is kinda in between TLS and no TLS where the first communications happen in plain text on the standard protocal port, the client (here the sender) will emit STARTTLS to let the server know, hey I support TLS, wanna switch to that? I think that was made to start using better security when servers could not yet enforce only TLS because it was not yet widely adopted...

These are just thoughs to try and help, feel free to correct me if I got anything wrong here, but I would start by making sure the mail server does not strictly enforce TLS 1.3 ONLY. This is very new and probably not widespread yet.

**Comment by thabaum** [comment]
I ran a full sweep of things and really buckled down on my environment as the issue.

My mail server logs helped me diagnose this issue further and in fact I found a thread from an issue similar to what I had that dated back to 2015 with me discussing the issue.

I was able to then get port 25 default port to send with SSL Enabled and I was unable to use 465 I would get the IO error.  SSL is unsupported anymore and so SSL Required option was on my server but the SMTP provider is actually starting a TLS connection so it IO errors.  

I removed SSL required on port 465 and set only allow secure authentication from SSL/TLS solved the problem.  I could send on 587 465 25 with or without authentication prior to sending on/off and with SSL enabled on DNN.

I went through the server logs and for some reason my server didn't like the newly created domain/user/password on the server so I recreated it's post office.  So the authentication was failing on port 587.  The IO error never made a log on my mail server, only a DNN error.

Maybe this information and solution will help someone like me in the future:

- The SMTP library will use STARTTLS to make an SSL connection.  
- You do NOT want SSL required on your mail server .
- You DO want to have SSL/TLS required for authentication on the mail server.
- DNN setting SSL enabled.
- good username/password to authenticate prior to sending or not.

Sorry it was a long time fooling with this over the years.  I can finally put this one to bed.  Thank you for the support!
